### PR TITLE
Multi server

### PR DESCRIPTION
**Pull request by connortechnology**
Mostly using the PR to manage breakage while merging all these other PR's...

### CODE REVIEW

File: distros/debian_cmake/apache.conf [hunk 1]
```diff
@@ -1,6 +1,6 @@
-Alias /zm /usr/share/zoneminder
+Alias /zm /usr/share/zoneminder/www
 
-<Directory /usr/share/zoneminder>
+<Directory /usr/share/zoneminder>/www
```
Comment (kylejohnson) [inline distros/debian_cmake/apache.conf]: typo here

**Comment by knnniggett** [comment]
I've got this built and running on CentOS 6 at the moment. 

When I navigate to Options -> Servers -> AddNewServer, attempt to add a new server, then click "save", nothing happens. The Servers tab remains empty.  

The following error in my Apache log appears relevant:
```
[Thu Nov 05 13:41:26 2015] [error] [client 127.0.0.1] PHP Notice:  Undefined index: Hostname in /usr/share/zoneminder/www/skins/classic/views/server.php on line 59, referer: https://localhost/zm/index.php?view=options&tab=servers
```

**Comment by knnniggett** [comment]
Just rebuilt this branch and performed the same test, adding a new host. The results were the same as the previous test, only the php error has changed:
```
[Wed Nov 18 08:38:45 2015] [error] [client 127.0.0.1] PHP Notice:  Undefined index: object in /usr/share/zoneminder/www/includes/actions.php on line 769, referer: https://localhost/zm/index.php?view=state&action=&apply=1&runState=stop&newState=
```

**Comment by connortechnology** [comment]
Helps if I merge master to it.  It currently works for me.

**Comment by knnniggett** [comment]
I've got this built and working on one server over the weekend. I ran out of time before I could create a second server and get the two to talk. Cleaning the house before Thanksgiving Day is a priority at the moment. 

I did have one question. Maybe this will become self evident as I continue to test, but how is authentication handled between the two servers when streaming?  Does each server need to have identical usernames and passwords created on each box?

**Comment by connortechnology** [comment]
authentication between the two servers?  The two servers don't really talk to each other.  They talk to the database.

You need a shared fs between them.  Live streaming has to come from the server that is doing the recording, but I don't think I've merged the code to alter the URL to the live feed yet.

File: web/api/app/Plugin/Crud [hunk 2]
```diff
@@ -1 +1 @@
-Subproject commit c3976f1478c681b0bbc132ec3a3e82c3984eeed5
+Subproject commit e22c1563a51d86aac0d5054beee28b4afb60c802
```
Comment (SteveGilvarry) [inline web/api/app/Plugin/Crud]: @connortechnology Something missed in a merge? c39... is the current Crud submodule version link.

**Comment by SteveGilvarry** [comment]
@connortechnology Sorry I was trying to send a PR against this to include some documentation changes, but they pushed straight to here. I assume caused by not coming from your own fork.
How far away from merge, seems to be OK and maybe we can do this weekend? Then we can flag the release and do some testing for 1.29.

**Comment by connortechnology** [comment]
I'll try to find some time to review this and clean up any messes later today.  

Not sure what you mean by pushed straight to here... I don't see any changes from you here in this PR..  But as far as I'm concerned, you can go ahead and push to this branch.  I trust you, and of course, git makes it easy to undo/blame.

File: src/zm_config.h.in [hunk 3]
```diff
@@ -65,6 +65,8 @@ struct StaticConfig
     std::string DB_USER;
     std::string DB_PASS;
     std::string PATH_WEB;
+	std::string SERVER_NAME;
+	unsigned int SERVER_ID;
```
Comment (kylejohnson) [inline src/zm_config.h.in:69]: Do we maybe want to make SERVER_ID an unsigned char?  256 should be plenty of server IDs, no?

File: src/zm_monitor.cpp [hunk 4]
```diff
@@ -394,11 +396,11 @@ Monitor::Monitor(
              + 64; /* Padding used to permit aligning the images buffer to 16 byte boundary */
 
     Debug( 1, "mem.size=%d", mem_size );
-	mem_ptr = NULL;
```
Comment (kylejohnson) [inline src/zm_monitor.cpp:397]: In the interest of keeping this huge PR as small and precise as possible, would you mind reverting the tab / spacing fixes?  Strictly, they aren't related to multi-server.

File: web/includes/control_functions.php [hunk 5]
```diff
@@ -20,9 +20,9 @@ function buildControlCommand( $monitor )
                 case 'focus' :
                 {
                     $factor = $_REQUEST['yge']/100;
-                    if ( $monitor['HasFocusSpeed'] )
```
Comment (kylejohnson) [inline web/includes/control_functions.php:23]: There are whole boat load of these changes (in this file and others).  What's the reason for them?

File: web/includes/functions.php [hunk 6]
```diff
@@ -94,6 +94,25 @@ function noCacheHeaders()
...
+	if ( isset( $_SERVER['HTTP_ORIGIN'] ) ) {
+
+		# The following is left for future reference/use.
+		$valid = false;
+		foreach( dbFetchAll( 'SELECT * FROM Servers' ) as $row ) {
+			$Server = new Server( $row );
+			if ( $_SERVER['HTTP_ORIGIN'] == $Server->Url() ) {
+				$valid = true;
+				header("Access-Control-Allow-Origin: " . $Server->Url() );
+				header("Access-Control-Allow-Headers: x-requested-with,x-request");
```
Comment (kylejohnson) [inline web/includes/functions.php:107]: Do we need multiple (looks like duplicate) `Access-Control-Allow-Headers` lines, or only a single one?
The `Access-Control-Allow-Origin` per-server makes sense, but I don't know that we need a matching `Access-Control-Allow-Headers` for each one.  Eh?

Comment (connortechnology) [inline web/includes/functions.php:107]: Not sure what you mean.  Those are different headers... note the Origin is where we are coming from and the Allow-Headers is what kind of request.  

I found that both were required, but that may change as other things change.

File: src/zm_config.h.in (hunk 3 above)
Comment (connortechnology) [inline src/zm_config.h.in:69]: Well, for the average person yeah... but what about when we take over the world and literally have millions of zm servers all over the internet to manage?

File: web/includes/functions.php (hunk 6 above)
Comment (kylejohnson) [inline web/includes/functions.php:107]: Sorry, I thought both headers were being added for each remote server, but I'm probably reading it wrong or misunderstanding how it works.  Think I follow now.

On December 21, 2015 12:52:04 PM EST, Isaac Connor notifications@github.com wrote:

> > @@ -94,6 +94,25 @@ function noCacheHeaders()
> >      header("Pragma: no-cache");         // HTTP/1.0
> >  }
> > 
> > +function CORSHeaders() {
> > -    if ( isset( $_SERVER['HTTP_ORIGIN'] ) ) {
> >   +
> > -        # The following is left for future reference/use.
> > -        $valid = false;
> > -        foreach( dbFetchAll( 'SELECT \* FROM Servers' ) as $row ) {
> > -            $Server = new Server( $row );
> > -            if ( $_SERVER['HTTP_ORIGIN'] == $Server->Url() ) {
> > -                $valid = true;
> > -                header("Access-Control-Allow-Origin: " . $Server->Url() );
> > -                header("Access-Control-Allow-Headers:
> >   x-requested-with,x-request");
> 
> Not sure what you mean.  Those are different headers... note the Origin
> is where we are coming from and the Allow-Headers is what kind of
> request.  
> 
> I found that both were required, but that may change as other things
> change.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/ZoneMinder/ZoneMinder/pull/1040/files#r48171031

## 

Sent from my Android device with K-9 Mail. Please excuse my brevity.

File: web/includes/control_functions.php (hunk 5 above)
Comment (SteveGilvarry) [inline web/includes/control_functions.php:23]: I knew this was an object thing and @connortechnology confirmed Monitor went from array to object.
https://github.com/ZoneMinder/ZoneMinder/pull/1040/files#diff-a63038416a07d6320efa13efef13405cR11

**Comment by SteveGilvarry** [comment]
Ready for Merge?
Edit and replace [ ] with [x]
- [ ] @connortechnology 
- [ ] @knnniggett 
- [ ] @kylejohnson 
- [x] @stevegilvarry

**Comment by knnniggett** [comment]
I'd like to build and test this one more time. 
I'll try to work this in over the Christmas holidays.

**Comment by kylejohnson** [comment]
I wouldn't have time to test this until after the new year, so I'll defer to @knnniggett's judgment.

**Comment by knnniggett** [comment]
I'm building this branch on a couple of machines now. Seems to be building fine.
Hopefully, I'll have a chance to test this tomorrow, before the in-laws arrive from out of town.

One thing I did notice was a lack of documentation stating what needs to be done, in addition to creating a server entry under Options and then assigning the server to the appropriate monitor(s). For instance, do I understand correctly that one must manually map the event storage from the remote server over to local?  Shouldn't we at least include some rudimentary documentation on how to set that part up?

**Comment by knnniggett** [comment]
@connortechnology 
When attempting to open a camera attached to a remote zoneminder server, the Monitor Id being used in the generated url is the local Monitor Id, rather than the Monitor Id of the camera on the remote server. Unless the two Monitor Id's are coincidentally the same, this prevents one from acquiring a stream and the usual "socket sendto" errors appear in the log.

For example, here is the correctly generated URL when opening my test camera from the server it is attached to:
`https://192.168.1.145/cgi-bin-zm/nph-zms?mode=jpeg&scale=100&maxfps=5&buffer=1000&monitor=1&connkey=898435&rand=1451241932`

And this is the generated URL when opening the same camera from the remote server:
`https://192.168.1.145/cgi-bin-zm/nph-zms?mode=jpeg&scale=100&maxfps=5&buffer=1000&monitor=2&connkey=669824&rand=1451242042`

Since the end user has no way to manage Monitor Id's in ZoneMinder, I would think it unreasonable to expect the Monitor Id's to match across servers in every case.

I'm trying to think of what a reasonable solution to this would be. Should we add another field to the monitor popup which allows the end user to select the remote monitor?

**Comment by SteveGilvarry** [comment]
@knnniggett all servers share the same database so MonitorID should be the same across all servers, and then server='blah' is used to control which monitors and other processes start on which server.

**Comment by knnniggett** [comment]
Maybe I am mistaken, but the goal as I understood it is to have multiple, independent ZoneMinder servers, with the potential to view all cameras from all servers from a single gui. We unify the frontend, but each backend works independently of one another. That includes the database.

Do you see any code that suggests otherwise? I have not, but it is hard to tell with all the commits in this pr.

**Comment by SteveGilvarry** [comment]
If they were independent you would have to setup duplicate copies of every Monitor on the Master server, and have the issue you are describing. Instead all the code now checks for server being itself before it runs anything. Easier to look at the files view for the changes.

And this https://github.com/ZoneMinder/ZoneMinder/pull/1040#issuecomment-158983814
Even put it into the documentation, what I didn't do was a specific setup instruction on multi-server. Or look at the shared file system for events, figured it was a more linux admin thing than zoneminder.

**Comment by knnniggett** [comment]
Oh right, got it. Guess I had information overload and didn't notice that earlier comment. 
I think a mini network topology diagram might help explain how to set up the underlying infrastructure. I'll draw one up.

**Comment by SteveGilvarry** [comment]
Should see if @pliablepixels has the original for the one he did to show all the components. We could extend that to show extra zmc servers. Also I think Storageareas removes the need for shared file share so 1.29 is an in between release, and then we go big release. I might try and find some time this week for a proposed branch merging a bunch of features. We shall call it Highlander 2 The Quickening lol

File: src/zm_monitor.cpp [hunk 7]
```diff
@@ -2706,197 +2692,143 @@ int Monitor::LoadFfmpegMonitors( const char *file, Monitor **&monitors, Purpose
...
+    int fps_report_interval = atoi(dbrow[col]); col++;
+    int ref_blend_perc = atoi(dbrow[col]); col++;
+    int alarm_ref_blend_perc = atoi(dbrow[col]); col++;
+    int track_motion = atoi(dbrow[col]); col++;
+
+    int signal_check_colour;
+    if ( dbrow[col][0] == '#' )
+        signal_check_colour = strtol(dbrow[col]+1,0,16);
+    else
+        signal_check_colour = strtol(dbrow[col],0,16);
```
Comment (SteveGilvarry) [inline src/zm_monitor.cpp:2788]: @connortechnology Did we lose a col++; here by accident?

**Comment by knnniggett** [comment]
Quoting Frankenstein, "It's Alive"!

Now that I understand how this works, a lot questions come to mind, but here are just a couple:
- Presumably, each ZoneMinder server is running its own instance of zmaudit, when we really only need one instance to run. Same thing might be true for filters as well. Is there an easy way to address that? I wonder if there might be side affects to having multiple instances of these services running.
- Since the database, and consequently the log, are now shared, we probably should do something to identify from which server the event originated from.

These are merely topics for discussion. I'm not arguing we need to address them before we merge this feature.

When setting up this new feature, I created the following network topology diagram to help me get a visual understanding of the system. I could build the entire system just with this diagram alone, but I've heard some feedback regarding the file format so I'll let my peers decide how to integrate this into our documentation:

zm_multiserver_topology

Here is a start on a procedural style README to step one through setting things up. The semantics used are the same as what is shown in the Network Topology diagram. What format will work best for this? Should I save it into the root folder as a README, add it in under the /docs folder, add it to readthedocs, something else?


New installs - DRAFT
============

1. Follow the normal instructions for your distro for installing ZoneMinder
   onto all the ZoneMinder Servers in the normal fashion. Mysql/Mariadb will
   only be needed on the machine targeted to be the Database Server.

2. On each ZoneMinder server, edit zm.conf. Find the ZM_DB_HOST variable and
   set it to the name or ip address of your Database Server.

3. Copy the file /usr/share/zoneminder/db/zm_create.sql from one of the
   ZoneMinder servers to the machine targeted as the Database Server.

4. Install mysql/mariadb server onto the Database Server.

5. It is advised to run "mysql_secure_installation" to help secure the server.

6. Using the password for the root account set during the previous step, create
   the ZoneMinder database and configure a database account for ZoneMinder to use:

     mysql -u root -p < zm_create.sql
     mysql -u root -p
        mysql> grant all
               on zm.* to 'zmuser'@'%' identified by 'zmpass';
        mysql> exit;
     mysqladmin -u root -p reload

   The database account credentials, zmuser/zmpass, are arbitrary. Set them to
   anything that suits your environment.

   Note that these commands are just an example and might not be secure enough
   for your environment.

7. If you have chosen to change the ZoneMinder database account credentials to
   something other than zmuser/zmpass, you must now update the zm.conf on each
   ZoneMinder Server. Change ZM_DB_USER and ZM_DB_PASS to the values you
   created in the previous step.

   Additionally, you must also edit 
   /usr/share/zoneminder/www/api/app/Config/database.php in a similar manner on
   each ZoneMinder Server. Scroll down and change login and password to the
   values you created in the previous step.

8. Open your browser and point it to the web console on any of the ZoneMinder
   Servers (they will all be the same). Open Options, click the Servers tab,
   and populate this screen with all of your ZoneMinder Servers.

9. When creating a new Monitor, remember to select the server the camera will
   be assigned to from the Server drop down box.


Upgrades
========

*** TBD ***

**Comment by connortechnology** [comment]
In regards to zmaudit, I find that generally for IO performance purposes, it is best if zmaudit only checks areas that are stored locally.  I have done some work to make zmaudit only look at local disks if desired, but that is not included here yet.  

For zmfilters, you can added a server selection to your filter to make it only look at events that were recorded on the specified or current server. So basically just the minimum support necessary to let people do what they want.

Agreed about the log.  I hadn't really thought about that, because I don't use the db log.  I will take a look.

**Comment by SteveGilvarry** [comment]
multiserver-2

@knnniggett did a new image with draw.io so we can store the xml for future editing by anyone.

As far as where we store the information, I think docs/installationguide as a separate multi-server guide, I will make a start on it and you can push changes to it

**Comment by knnniggett** [comment]
I don't agree with the way you've drawn the diagram, but if you want me to change it in draw.io I can do that after you push it up.

**Comment by knnniggett** [comment]
Alright, so it seems the last remaining thing is Documentation, which is being addressed in #1211 .
Are we ready merge to merge this PR? I think we are.
I'll merge it tomorrow morning on the 1st at the latest if I don't hear from anyone.

**Comment by knnniggett** [comment]
Here is the what @SteveGilvarry and I settled on for the diagram:

zm-multiserver

**Comment by knnniggett** [comment]
Whew. It is finished.

**Comment by pliablepixels** [comment]
you took out the mobile client image!

**Comment by knnniggett** [comment]
Mobile clients specifically are not relevant to installing the underlying infrastructure for mutli-server, which is what the diagram is for. Instead, there is one symbol, labelled "Video Client (Typical)" which represents any form of video client. The emphasis is on which network a typical client device should be attached to, rather than the type of the client device.

We can create additional diagrams, which show a more general installation of ZoneMinder, if we need to.

**Comment by SteveGilvarry** [comment]
Offers @pliablepixels a shoulder to cry on.

**Comment by pliablepixels** [comment]
In knnniggett's world, a video client (typical) looks like an HP Compaq desktop :-)
I need to infuse him with some bling, "yo!". Mobile is the new normal.

**Pull request merged**